File: core/core-web/src/test/java/ch/difty/scipamato/core/web/paper/search/SearchOrderChangeEventTest.java
package ch.difty.scipamato.core.web.paper.search;

import static org.assertj.core.api.Assertions.assertThat;

import nl.jqno.equalsverifier.EqualsVerifier;
import nl.jqno.equalsverifier.Warning;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import ch.difty.scipamato.core.web.paper.SearchOrderChangeEvent;

@ExtendWith(MockitoExtension.class)
class SearchOrderChangeEventTest {

    private SearchOrderChangeEvent e;

    @Mock
    private AjaxRequestTarget targetMock, targetMock2;

    @Test
    void canRetrieveTarget() {
        e = new SearchOrderChangeEvent(targetMock);
        assertThat(e.getTarget()).isEqualTo(targetMock);
    }

    @Test
    void usingMinimalConstructor_doesNotSetAnySpecialStuff() {
        e = new SearchOrderChangeEvent(targetMock);
        assertThat(e.getExcludedId()).isNull();
        assertThat(e.getDroppedConditionId()).isNull();
        assertThat(e.isNewSearchOrderRequested()).isFalse();
    }

    @Test
    void usingWithExcludedPaperId_doesAddExclusionId() {
        e = new SearchOrderChangeEvent(targetMock).withExcludedPaperId(5L);
        assertThat(e.getExcludedId()).isEqualTo(5L);
    }

    @Test
    void usingWithDroppedConditionId_doesAddConditionId() {
        e = new SearchOrderChangeEvent(targetMock).withDroppedConditionId(5L);
        assertThat(e.getDroppedConditionId()).isEqualTo(5L);
    }

    @Test
    void requestingNewSearchOrder_setsFlagAccordingly() {
        e = new SearchOrderChangeEvent(targetMock).requestingNewSearchOrder();
        assertThat(e.isNewSearchOrderRequested()).isTrue();
    }

    @Test
    void requestingNewSearchOrder_withExcludedPaperIdAndNewSearchOrderRequest_newSearchOrderRequestWins() {
        e = new SearchOrderChangeEvent(targetMock)
            .withExcludedPaperId(5L)
            .requestingNewSearchOrder();
        assertThat(e.isNewSearchOrderRequested()).isTrue();
        assertThat(e.getExcludedId()).isNull();
    }

    @Test
    void requestingNewSearchOrder_withNewSearchOrderRequestAndThenExcludedPaperId_exclusionWins() {
        e = new SearchOrderChangeEvent(targetMock)
            .requestingNewSearchOrder()
            .withExcludedPaperId(5L);
        assertThat(e.getExcludedId()).isEqualTo(5L);
        assertThat(e.isNewSearchOrderRequested()).isFalse();
    }

    @Test
    void canOverrideTarget() {
        e = new SearchOrderChangeEvent(targetMock);
        assertThat(e.getTarget()).isEqualTo(targetMock);
        e.setTarget(targetMock2);
        assertThat(e.getTarget()).isEqualTo(targetMock2);
    }

    @Test
    void equals() {
        EqualsVerifier
            .forClass(SearchOrderChangeEvent.class)
            .withRedefinedSuperclass()
            .suppress(Warning.STRICT_INHERITANCE, Warning.NONFINAL_FIELDS)
            .verify();
    }
}

File: core/core-web/src/test/kotlin/ch/difty/scipamato/core/web/paper/search/SearchOrderChangeEventTest.kt
package ch.difty.scipamato.core.web.paper.search

import ch.difty.scipamato.common.AjaxRequestTargetSpy
import ch.difty.scipamato.core.web.paper.SearchOrderChangeEvent
import nl.jqno.equalsverifier.EqualsVerifier
import nl.jqno.equalsverifier.Warning
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeNull
import org.amshove.kluent.shouldBeTrue
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test

internal class SearchOrderChangeEventTest {

    private lateinit var e: SearchOrderChangeEvent

    private val targetDummy = AjaxRequestTargetSpy()

    @AfterEach
    fun tearDown() {
        targetDummy.reset()
    }

    @Test
    fun canRetrieveTarget() {
        e = SearchOrderChangeEvent(targetDummy)
        e.target shouldBeEqualTo targetDummy
    }

    @Test
    fun usingMinimalConstructor_doesNotSetAnySpecialStuff() {
        e = SearchOrderChangeEvent(targetDummy)
        e.excludedId.shouldBeNull()
        e.droppedConditionId.shouldBeNull()
        e.isNewSearchOrderRequested.shouldBeFalse()
    }

    @Test
    fun usingWithExcludedPaperId_doesAddExclusionId() {
        e = SearchOrderChangeEvent(targetDummy).withExcludedPaperId(5L)
        e.excludedId shouldBeEqualTo 5L
    }

    @Test
    fun usingWithDroppedConditionId_doesAddConditionId() {
        e = SearchOrderChangeEvent(targetDummy).withDroppedConditionId(5L)
        e.droppedConditionId shouldBeEqualTo 5L
    }

    @Test
    fun requestingNewSearchOrder_setsFlagAccordingly() {
        e = SearchOrderChangeEvent(targetDummy).requestingNewSearchOrder()
        e.isNewSearchOrderRequested.shouldBeTrue()
    }

    @Test
    fun requestingNewSearchOrder_withExcludedPaperIdAndNewSearchOrderRequest_newSearchOrderRequestWins() {
        e = SearchOrderChangeEvent(targetDummy).withExcludedPaperId(5L).requestingNewSearchOrder()
        e.isNewSearchOrderRequested.shouldBeTrue()
        e.excludedId.shouldBeNull()
    }

    @Test
    fun requestingNewSearchOrder_withNewSearchOrderRequestAndThenExcludedPaperId_exclusionWins() {
        e = SearchOrderChangeEvent(targetDummy).requestingNewSearchOrder().withExcludedPaperId(5L)
        e.excludedId shouldBeEqualTo 5L
        e.isNewSearchOrderRequested.shouldBeFalse()
    }

    @Test
    fun canOverrideTarget() {
        e = SearchOrderChangeEvent(targetDummy)
        e.target shouldBeEqualTo targetDummy

        val targetDummy2 = AjaxRequestTargetSpy()
        e.target = targetDummy2
        e.target shouldBeEqualTo targetDummy2
    }

    @Test
    fun equals() {
        EqualsVerifier
            .forClass(SearchOrderChangeEvent::class.java)
            .withRedefinedSuperclass()
            .suppress(Warning.STRICT_INHERITANCE, Warning.NONFINAL_FIELDS)
            .verify()
    }
}
